Chaincase.UI/Pages/CoinJoinPage.razor

```diff
@@ -1,51 +1,142 @@
+        @*Coin Selection <ion-icon name="chevron-forward-outline" />*@
+        <SelectCoinsModal @bind-Visible="@IsSelectCoinsModalVisible"
+                          SelectCoinsViewModel="@ViewModel.CoinList" />
+        @* Progress Bar *@
+        <ion-text color="primary">
+            <h5>Percentage of peers queued in most advanced round </h5>
+        </ion-text>
+        <ion-progress-bar value="@ViewModel.QueuedPercentage" />
+        @* List Coin Join round info**@
+
+        <ul class="list-group">
+            <li class="badge">
+                ₿ Required to CoinJoin
+                <span class="primary">@ViewModel.RequiredBTC</span>
+            </li>
+            <li class="badge">
+                Phase
+                <span class="success"> @ViewModel.RoundPhaseState.Phase </span>
+            </li>
+            <li class="badge">
+                Registration Ends in
+                <span class="primary">@ViewModel.TimeLeftTillRoundTimeout </span>
+            </li>
+            <li class="badge">
+                Joining With at Least
+                <span class="primary">@ViewModel.PeersNeeded peers</span>
+            </li>
+            <li class="badge">
+                Registered Peers
+                <span class="primary"> @ViewModel.PeersRegistered</span>
+            </li>
+            <li class="badge">
+                Coordinator Fee
+                <span class="primary">@ViewModel.CoordinatorFeePercent % per anonymity set </span>
+            </li>
+        </ul>
+        <div>
+            <IonPasswordInput @bind-Value="password" placeholder="Password" />
+        </div>
+        @if (ViewModel.IsQueuedToCoinJoin)
+        {
+            <ion-button @onclick='ExitCoinJoin' @onclick:preventDefault="true" expand="block">Cancel </ion-button>
+        }
+        else
+        {
+            <ion-button @onclick='CoinJoin' @onclick:preventDefault="true" expand="block">COINJOIN</ion-button>
+            <ion-button @onclick='ExitCoinJoin' @onclick:preventDefault="true" expand="block">Cancel </ion-button>
+        }
+        <IonToast @bind-Visible="IsToastVisible" @bind-message="ToastMessage" showCloseButton="true" duration="2000"></IonToast>
+        }
 </ion-content>
 
 @code {
+    private bool IsSelectCoinsModalVisible = false;
+    private bool IsToastVisible = false;
+    private string ToastMessage = "";
+    private const int ToastDuration = 2000; // ms
 
-    bool IsQueuedToCoinJoin = true;
-    string RequiredBTC = "0.01";
 
+    private string password = "";
+    public CoinJoinPage()
+    {
+    }
 
     protected override void OnInitialized()
     {
-        UiStateService.Title = "SEND";
+        ViewModel = CoinJoinViewModel;
+        UiStateService.Title = "COIN JOIN";
     }
 
     void ExitCoinJoin()
     {
-        return;
+        NavigationManager.NavigateTo("/");
     }
 
     void CoinJoin()
     {
-        return;
+        try
+        {
+            ViewModel.JoinRound(password);
```

Comment:
This blocks the UI thread for too long. Consider await

Reply:
As far as I can tell await still blocks the UI thread making it unclear still how to e.g. set loading = true, render, wait for the coinjoin task' set loading = false, render and continue. We might have to set coinjoin to a separate event handler that gets triggered. the trigger sets loading/busy = true then gets rendered, then the handler shuts loading off.